update DOS page (#3451)
dfinity/portal: docs/developer-docs/security/security-best-practices/dos.mdx

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Security Best Practices: Denial of Service

<MarkdownChipRow labels={["Intermediate", "Security", "Concept"]} />

## Protect Against DoS and DDoS Attacks

### Security Concern

A Denial of Service (DoS) attack aims to make a system unavailable by overwhelming it with requests or data. A Distributed Denial of Service (DDoS) attack is a more sophisticated version, where the attack originates from multiple sources, making it harder to block. An attacker will typically search for operations that are free to be executed by anyone but which are expensive for the application in terms of certain resources such as storage, memory usage, network bandwidth, computing resources, etc. In the case of canisters, such attacks can aim to deplete cycles, making the canister unable to process legitimate requests. The reverse gas model means that a dapp needs to implement strategies to deal with this.

### Recommendation

To protect your canisters from DoS and DDoS attacks, consider the following strategies:

> * **Bot prevention techniques**: Use methods like captchas or proof of work to ensure only legitimate users can access your canister. Captchas help verify that the user is human, while proof of work requires the user to spend computational resources to proceed, deterring automated attacks. An example implementation can be found in the [Internet Identity project](https://github.com/dfinity/internet-identity).

Reviewer: Does II also implement PoW or only Captcha?
Could you link the piece of code that does this?

* **Monitor cycles usage**: Regularly track your canisters cycles consumption and set alerts for any sudden spikes that may indicate an attack.

> * **Ingress message charging**: While charging for ingress messages (external requests to the canister) is not natively supported, custom solutions could be implemented to make sure that any expensive actions have costs associated with them.

Reviewer: ```suggestion
* **Ingress message charging**: While charging for ingress messages (external requests to the canister) is not natively supported, custom solutions could be implemented to make sure that any expensive actions have costs associated with them.
* **Filter ingress messages using inspect message**:  Certain non-critical checks can be placed in the inspect message function to filter out ingress update messages before they are executed by all nodes of a subnet. Since this code only runs on a single node, the execution does not consume cycles but it also shouldn't be relied upon for security critical checks such as access control. However, they can efficiently reject certain ingress messages early. Read the corresponding [documentation](https://internetcomputer.org/docs/current/references/ic-interface-spec#system-api-inspect-message) and [security best practice](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/iam#do-not-rely-on-ingress-message-inspection) carefully for the caveats.
```

Learn more about managing memory and compute resources in our [Storage](https://internetcomputer.org/docs/current/developer-docs/gas-cost#storage) and [Compute](https://internetcomputer.org/docs/current/developer-docs/gas-cost#execution) guides.
* **Subnet and canister distribution**: Implement a smart canister deployment strategy by monitoring the load on subnets. You can choose to deploy new canisters on less busy subnets or adopt a multi-canister architecture that balances the load across subnets. Be mindful to minimize inter-subnet communication for canisters that frequently interact with each other. Additionally, avoid deploying to known high-traffic subnets where possible, though keep in mind that resource usage can change unexpectedly with new dapps.

## Handle Expensive Calls

### Security Concern

> Some calls (update or query) might be expensive in terms of the memory or cycles they consume, making it easy for malicious users to render your canister unusable by draining resources.

Reviewer: ```suggestion
Some calls (update or query) might be expensive in terms of the memory or cycles they consume. For example, any function using chain-key signing or HTTPS outcalls are relatively expensive. See the [additional documentation](https://internetcomputer.org/docs/current/developer-docs/gas-cost#special-features) on cycles cost details and for other examples.

An attacker will target expensive calls to drain cycles balance or available memory quickly.
```

### Recommendation

> * **Use captchas or proof of work**: If your application provides expensive operations (such as complex computations or large data storage), consider requiring users to complete a captcha or proof of work before proceeding. This can help prevent bots from abusing your system.

Reviewer: ```suggestion
* **Use captchas**: Expensive operations should require a captcha to be solved. Try to use a library to implement a captcha instead of a cloud service as such a service would require HTTPS outcalls and isn't decentralized.
* **Use PoW (proof-of-work)**: Require a proof-of-work challenge to be solved by the client for any expensive operation. The parameters need to be carefully chosen to require sufficient computation per call to the expensive operation without creating too much impact for legitimate clients. Don't forget to consider clients on slow and older mobile devices whilst protecting against attackers on modern multi-GPU systems. Certain algorithms can limit the performance increase of GPUs to improve this uneven battlefield.
```

* **Charge for expensive calls**: You can require that certain expensive calls from other canisters include cycles to compensate for the resources consumed. One can charge for ingress messages as well. However, that is not currently supported by the protocol itself and a custom solution, such as pre-paying a certain amount, would need to be designed.
* **Differentiate between update and query calls**: Expensive computations should generally be avoided for update calls unless absolutely necessary. While query calls are not authenticated, they are faster and less resource-intensive. To check whether a method was called as a query or update, you can use `ic0.in_replicated_execution()`.

### Further Recommendations

> - Monitor cycles consumption regularly.

Reviewer: ```suggestion
- Automatically monitor cycles consumption and set appropriate alerts for cycles consumption rate and balance.
```

- Implement early authentication and rate limiting for your canisters.

> - Be aware of attacks targeting high cycles-consuming calls.

Reviewer: - See the "Cycle balance drain attacks section" in [how to audit an ICP canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister).

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

> # Security Best Practices: Denial of Service

Reviewer 2: ```suggestion
# Security best practices: Denial of service
```

<MarkdownChipRow labels={["Intermediate", "Security", "Concept"]} />

> ## Protect Against DoS and DDoS Attacks

Reviewer 2: ```suggestion
## Protect against DoS and DDoS attacks
```

> ### Security Concern

Reviewer 2: ```suggestion
### Security concern
```

* **Monitor cycles usage**: Regularly track your canisters cycles consumption and set alerts for any sudden spikes that may indicate an attack.
* **Ingress message charging**: While charging for ingress messages (external requests to the canister) is not natively supported, custom solutions could be implemented to make sure that any expensive actions have costs associated with them.
* **Filter ingress messages using inspect message**:  Certain non-critical checks can be placed in the inspect message function to filter out ingress update messages before they are executed by all nodes of a subnet. Since this code only runs on a single node, the execution does not consume cycles but it also shouldn't be relied upon for security critical checks such as access control. However, they can efficiently reject certain ingress messages early. Read the corresponding [documentation](https://internetcomputer.org/docs/current/references/ic-interface-spec#system-api-inspect-message) and [security best practice](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/iam#do-not-rely-on-ingress-message-inspection) carefully for the caveats.

> ## Protect Against Noisy Neighbors

Reviewer 2: ```suggestion
## Protect against noisy neighbors
```

> ### Security Concern

Reviewer 2: ```suggestion
### Security concern
```

In a shared resource environment like the Internet Computer, multiple canisters can run on the same subnet. If one canister consumes too many resources (CPU, memory, etc.), it can negatively impact the performance of others on the same subnet. This is known as the "noisy neighbor" problem.

### Recommendation

To mitigate the "noisy neighbor" issue, manage your canister's resource allocation effectively:

> * **Memory allocation**: Memory can be reserved per canister by setting `memory_allocation`. This ensures that your canister can always allocate memory up to the requested `memory_allocation`. Thereby preventing other canisters from using up the subnets available memory. Note that this also reduces the upper bound of memory the canister can allocate to the same value. Monitoring actual memory usage against this value is important to avoid availability issues.

Reviewer 2: ```suggestion
* **Memory allocation**: Memory can be reserved per canister by setting `memory_allocation`, ensuring that your canister can always allocate memory up to the requested `memory_allocation` and preventing other canisters from using up the subnets available memory. Note that this also reduces the upper bound of memory the canister can allocate to the same value. Monitoring actual memory usage against this value is important to avoid availability issues.
```

> * **Compute reservation**: Similar to memory, computing power can also be reserved by setting `compute_allocation` to a value between 0 and 100 which denotes the percentage of one CPU core to be reserved for this canister. A value of 50 means that every 2 rounds, the canister will be scheduled to execute a message. This guarantees a minimal progress you canister can make which protects against noisy neighbours. Both allocations are reserving resources for your canister on the subnet which prevents the other canisters from using them. Hence, they come at a cost. Memory allocation is charged as if all that memory would be allocated. Compute allocation is currently charged at 10M cycles per percentage point.

Reviewer 2: ```suggestion
* **Compute reservation**: Similar to memory, computing power can also be reserved by setting `compute_allocation` to a value between 0 and 100, which denotes the percentage of one CPU core to be reserved for this canister. A value of 50 means that every 2 rounds, the canister will be scheduled to execute a message. This guarantees the minimal progress you canister can make, which protects against noisy neighbors. Both allocations are reserving resources for your canister on the subnet, which prevents the other canisters from using them. Hence, they come at a cost. Memory allocation is charged as if all that memory would be allocated. Compute allocation is currently charged at 10M cycles per percentage point.
```

> Learn more about managing memory and compute resources in our [Storage](https://internetcomputer.org/docs/current/developer-docs/gas-cost#storage) and [Compute](https://internetcomputer.org/docs/current/developer-docs/gas-cost#execution) guides.

Reviewer 2: ```suggestion
Learn more about managing memory and compute resources in the [Storage](https://internetcomputer.org/docs/current/developer-docs/gas-cost#storage) and [Compute](https://internetcomputer.org/docs/current/developer-docs/gas-cost#execution) guides.
```

:::info
When the subnet grows above 450GiB, then the new reservation mechanism activates. Every time a canister allocates new storage bytes, the system sets aside some amount of cycles from the main balance of the canister. These reserved cycles will be used to cover future payments for the newly allocated bytes. The reserved cycles are not transferable and the amount of reserved cycles depends on how full the subnet is. For example, it may cover days, months, or even years of payments for the newly allocated bytes. It is important to note that the reservation mechanism applies only to the newly allocated bytes and does not apply to the storage already in use by the canister. See more at [Reseourse reservations](https://forum.dfinity.org/t/increasing-subnet-storage-capacity-and-introducing-resource-reservation-mechanism/23447).
:::

> ## Handle Expensive Calls

Reviewer 2: ```suggestion
## Handle expensive calls
```

> ### Security Concern

Reviewer 2: ```suggestion
### Security concern
```

Some calls (update or query) might be expensive in terms of the memory or cycles they consume. For example, any function using chain-key signing or HTTPS outcalls are relatively expensive. See the [additional documentation](https://internetcomputer.org/docs/current/developer-docs/gas-cost#special-features) on cycles cost details and for other examples.

An attacker will target expensive calls to drain cycles balance or available memory quickly.

### Recommendation

> * **Use captchas**: Expensive operations should require a captcha to be solved. Try to use a library to implement a captcha instead of a cloud service as such a service would require HTTPS outcalls and isn't decentralized.
> * **Use PoW (proof-of-work)**: Require a proof-of-work challenge to be solved by the client for any expensive operation. The parameters need to be carefully chosen to require sufficient computation per call to the expensive operation without creating too much impact for legitimate clients. Don't forget to consider clients on slow and older mobile devices whilst protecting against attackers on modern multi-GPU systems. Certain algorithms can limit the performance increase of GPUs to improve this uneven battlefield.
> * **Charge for expensive calls**: You can require that certain expensive calls from other canisters include cycles to compensate for the resources consumed. One can charge for ingress messages as well. However, that is not currently supported by the protocol itself and a custom solution, such as pre-paying a certain amount, would need to be designed.
> * **Differentiate between update and query calls**: Expensive computations should generally be avoided for update calls unless absolutely necessary. While query calls are not authenticated, they are faster and less resource-intensive. To check whether a method was called as a query or update call, you can use `ic0.in_replicated_execution()`.

Reviewer 2: ```suggestion
* **Use captchas**: Expensive operations should require a captcha to be solved. Try to use a library to implement a captcha instead of a cloud service, as such a service would require HTTPS outcalls and isn't decentralized.
* **Use PoW (proof-of-work)**: Require a proof-of-work challenge to be solved by the client for any expensive operation. The parameters need to be carefully chosen to require sufficient computation per call to the expensive operation without creating too much impact for legitimate clients. Don't forget to consider clients on slow and older mobile devices whilst protecting against attackers on modern multi-GPU systems. Certain algorithms can limit the performance increase of GPUs to improve this uneven battlefield.
* **Charge for expensive calls**: You can require that certain expensive calls from other canisters include cycles to compensate for the resources consumed. In addition, one can charge for ingress messages. However, that is not currently supported by the protocol itself, and a custom solution, such as pre-paying a certain amount, would need to be designed.
* **Differentiate between update and query calls**: Expensive computations should generally be avoided for update calls unless absolutely necessary. While query calls are not authenticated, they are faster and less resource-intensive. To check whether a method was called as a query or update call, you can use `ic0.in_replicated_execution()`.
```

> ### Further Recommendations

Reviewer 2: ```suggestion
### Further recommendations
```

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Security Best Practices: Denial of Service

<MarkdownChipRow labels={["Intermediate", "Security", "Concept"]} />

## Introduction

Canisters on the Internet Computer are responsible for managing their own resources, including paying for execution, storage, and messaging with cycles. This makes them efficient, but also exposes them to risks such as Denial of Service (DoS) attacks, which can drain their cycles, and the "noisy neighbor" problem, where other canisters consume excessive resources, impacting performance.

In this guide, we will explore how to protect your canisters from these risks and manage resources effectively.

> ## 1. Protecting Against DoS and DDoS Attacks

Reviewer: I'd move this paragraph into the introduction so the rest can become "best practices".

### Understanding DoS and DDoS Attacks

A Denial of Service (DoS) attack aims to make a system unavailable by overwhelming it with requests or data. A Distributed Denial of Service (DDoS) attack is a more sophisticated version, where the attack originates from multiple sources, making it harder to block. 

> In the case of canisters, such attacks can quickly deplete cycles, making the canister unable to process legitimate requests. Learn more about DoS in the [Security Best Practices: Denial of Service](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/dos) guide.

Author: removed

> In the case of canisters, such attacks can quickly deplete cycles, making the canister unable to process legitimate requests. Learn more about DoS in the [Security Best Practices: Denial of Service](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/dos) guide.

Reviewer: I believe I'm still seeing the reference to https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/dos in the latest commit.

### Preventing DoS and DDoS Attacks

To protect your canisters from DoS and DDoS attacks, consider the following strategies:

> - **Bot prevention techniques**: Use methods like captchas or proof of work to ensure only legitimate users can access your canister. Captchas help verify that the user is human, while proof of work requires the user to spend computational resources to proceed, deterring automated attacks. An example implementation can be found in the [Internet Identity project](https://github.com/dfinity/internet-identity).

Reviewer: Add a reference to the relevant code.

Author: fixed

Author: fixed

- **Monitor cycles usage**: Regularly track your canisters cycles consumption and set alerts for any sudden spikes that may indicate an attack.

> - **Ingress message charging**: While charging for ingress messages (external requests to the canister) is not natively supported, custom solutions could be implemented to make sure that any expensive actions have costs associated with them.

Reviewer: What would that look like?
Requiring an approval on a ledger where the dapp can charge the user from?

Reviewer 3: It's similar the wallet (proxy) canister idea we use with dfx.

Reviewer 3: It's similar the wallet (proxy) canister idea we use with dfx.

Reviewer: I'm not 100% sure I understand what you mean. Would you mind writing a suggested recommendation in this PR?

## 2. Managing Resource Allocation

> ### The "Noisy Neighbor" Problem

Reviewer: Another strategy to deal with this problem is to have an efficient migration strategy to another subnet or at least have a multi canister architecture where you take into account the load on each subnet to decide where to deploy new canisters. Of course also taking into account reducing inter subnet calls as much as possible for canisters that talk a lot to each other.

Some dapps specifically avoid some of the known busy subnets or just target a subnet where current resource consumption is very low. Of course, you never know when some noisy dapp will deploy onto your subnet.

Author: agreed

Author: added

In a shared resource environment like the Internet Computer, multiple canisters can run on the same subnet. If one canister consumes too many resources (CPU, memory, etc.), it can negatively impact the performance of others on the same subnet. This is known as the "noisy neighbor" problem.

### Protecting Against Noisy Neighbors

To mitigate the "noisy neighbor" issue, manage your canister's resource allocation effectively:

> - **Memory allocation**: You can set a memory reservation for your canister through the `memory_allocation` setting, ensuring that your canister has enough reserved memory to operate even if other canisters consume large amounts of resources.

Reviewer 4: > enough reserved memory

it might be worthwhile pointing out that while the memory allocation guarantees a canister to have the specified amount of memory available, it also acts as an upper bound on the actual memory usage and messages will trap if they exceeded the specified memory allocation

> - **Memory allocation**: You can set a memory reservation for your canister through the `memory_allocation` setting, ensuring that your canister has enough reserved memory to operate even if other canisters consume large amounts of resources.

Reviewer: ```suggestion
- **Memory allocation**: Memory can be reserved per canister by setting `memory_allocation`. This ensures that your canister can always allocate memory up to the requested memory_allocation. This prevents other canisters from using up the subnets available memory. Note that this also reduces the upper bound of memory the canister can allocate to the same value. Monitoring actual memory usage against this value is important to avoid availability issues.
```

> - **Compute reservation**: Similar to memory, you can also reserve compute power for your canister, ensuring it has enough processing power to handle requests. Learn more about managing memory and compute resources in our [Storage](https://internetcomputer.org/docs/current/developer-docs/gas-cost#storage) and [Compute](https://internetcomputer.org/docs/current/developer-docs/gas-cost#compute) guides.

Reviewer: https://internetcomputer.org/docs/current/developer-docs/gas-cost#compute doesn't seem to work as there is no section "compute" on this page.

Author: fixed

> - **Compute reservation**: Similar to memory, you can also reserve compute power for your canister, ensuring it has enough processing power to handle requests. Learn more about managing memory and compute resources in our [Storage](https://internetcomputer.org/docs/current/developer-docs/gas-cost#storage) and [Compute](https://internetcomputer.org/docs/current/developer-docs/gas-cost#compute) guides.

Reviewer: ```suggestion
- **Compute reservation**: Similar to memory, you can also reserve computing power for your canister, ensuring it has enough processing power to handle requests. Learn more about managing memory and compute resources in our [Storage]

Both allocations are reserving resources for your canister on the subnet which prevents the other canisters from using them. Hence, they come at a cost. Memory allocation is charged as if all that memory would be allocated. Compute allocation is currently charged at 10M cycles per percentage point.

(https://internetcomputer.org/docs/current/developer-docs/gas-cost#storage) and [Compute](https://internetcomputer.org/docs/current/developer-docs/gas-cost#compute) guides.
```

Author: fixed

## 3. Managing Cycles

### Cycles and Reserved Cycles

> Cycles are a key part of the Internet Computer's resource management system, as they are used to pay for canister operations. To prevent your canister from running out of cycles unexpectedly, you can reserve cycles specifically for future use, helping to ensure consistent operation.

Reviewer 5: I think this is a misunderstanding of what "reserved" cycles are. Your canister cannot reserve cycles for future use -- it's actually kinda the opposite. The canister can tell the system, if you need to reserve some cycles due to memory allocations on a busy subnet (this is the only case where we "reserve" cycles from the canister), then you cannot "reserve" more than I'm specifying in my canister's settings (i.e. memory allocations made by the canister that would need more than specified in "reserved" would fail).

Reviewer 4: https://github.com/dfinity/portal/pull/3451#discussion_r1753355620

Reviewer 5: Oh I missed that, there were lots of comments already before I joined the party.

### Managing Reserved Cycles

> It's important to monitor and adjust your reserved cycles regularly to meet your canister's needs. Learn more about managing cycles in our [Cycle Bookkeeping and Resource Consumption](https://internetcomputer.org/docs/current/references/ic-interface-spec#cycle-bookkeeping-and-resource-consumption) guide.

Reviewer 4: ```suggestion
It's important to monitor and adjust your cycles regularly to meet your canister's needs. Learn more about managing cycles in our [Cycle Bookkeeping and Resource Consumption](https://internetcomputer.org/docs/current/references/ic-interface-spec#cycle-bookkeeping-and-resource-consumption) guide.
```
"reserved" cycles might be misunderstood as the cycles set aside from the main balance if the canister makes memory allocations on a subnet that is already short on memory; if we want to go into the details of "reserved" cycles, a separate paragraph might be more suitable

## 4. Handling Expensive Calls

### Security Concern: Expensive Calls and DoS

Some calls (update or query) might be expensive in terms of the memory or cycles they consume, making it easy for malicious users to render your canister unusable by draining resources.

> ### Mitigating the Risk

Reviewer 3: It would be worthwhile adding about the new candid Rust library features here (`decoding_quota` and `skipping_quota`) which are also a main expensive call risk. 

cc: @mraszyk

Reviewer 6: @venkkatesh-sekar it might be much to ask from Diego for this PR, as he might not have the context? Maybe consider this for a follow-up PR?

Reviewer 3: I agree, I will create a backlog to follow up on this.

- **Use captchas or proof of work**: If your application provides expensive operations (such as complex computations or large data storage), consider requiring users to complete a captcha or proof of work before proceeding. This can help prevent bots from abusing your system.

> - **Charge for expensive calls**: You can require that certain expensive calls from other canisters include cycles to compensate for the resources consumed. 

Author: Agreed

> - **Charge for expensive calls**: You can require that certain expensive calls from other canisters include cycles to compensate for the resources consumed. 

Author: added

> - **Differentiate between update and query calls**: Expensive computations should generally be avoided for update calls unless absolutely necessary. While query calls are not authenticated, they are faster and less resource-intensive. To check whether a method was called as a query or update, you can use `ic0.data_certificate_present()`.

Author: done
